gh-113804: Support formatting floats in hexadecimal notation

skirpichev opened the pull request:
The ``x`` and ``X`` format types (in ``str.format()``, f-strings and old-style string formatting) now support also formatting floats as a hexadecimal string, i.e. in the style ``[±][0x]h[.hhh]p±d`` (which is essentially a base-16 scientific notation, but with base-2 exponent, written in decimal).

``_Py_dg_dtoa_hex()`` helper is based on the ``float.hex()`` method with additional support for the precision setting and some format flags.  Note (c.f. printf's ``'a'`` format type of the C stdlib), that ``'#'`` option is used to control the *prefix* (off by default).

Trailing zeros (and the dot) in the fractional part are excluded.  (This change doesn't affect the ``float.hex()`` method and ``float.fromhex()``/``float.hex()`` round-trip.)

Examples:
```pycon
>>> f'{-0.1:#x}'
'-0x1.999999999999ap-4'
>>> (-0.1).hex()
'-0x1.999999999999ap-4'
>>> f'{3.14159:+#X}'
'+0X1.921F9F01B866EP+1'
>>> f'{3.14159:.3x}'
'1.922p+1'
```

Minor changes:
* added ``Py_hexdigits_upper`` constant
* tests for RaisingNumber are moved (to test also bytes)

------------------

Notes
-----

Implementation was changed to use "x"/"X" format types per suggestions both in the [discussion thread](https://discuss.python.org/t/support-hexadecimal-floating-point-literals/41848) and here.  This (1) makes output slightly more compact per default and (2) allows to extend formatting support to [old-style string formatting](https://docs.python.org/3/library/stdtypes.html#old-string-formatting) (using "a" letter will conflict with existing ``ascii()`` converter).

Personally I would choose "a" format type (like printf) instead and keep the prefix mandatory, to simplify interaction with other languages (e.g. copy-pasting output to C code as a hexadecimal literal) or within in the CPython repl itself, if #114668 will be accepted.  It worth to note, that both the gmpy2 and the bigfloat packages use the "a" format type.

Open issues:
- Should we add also "a" format type, with C-like behaviour, for ``str.format()``/f-strings to be more compatibile with existing libraries (e.g. gmpy2)?
- ~~Should we change at all old-style string formatting?~~
- Should we support binary notation (using e.g. "b"/"B" format types as the MPFR does)?
- Should we adjust tests to use ``f"{v:#x}"`` instead of ``v.hex()``?
- Should we deprecate (at least soft deprecate) ``float.hex()`` method?
- Related pr: #114668 (require a PEP)

--------------------------------------------------------------------

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-113804 -->
* Issue: gh-113804
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--113805.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

skirpichev committed:
gh-113804: Support "x" and "X" format types for floats

_Py_dg_dtoa_hex() helper is based on float.hex() with additional support
for the precision setting and some format flags.  Note (c.f. ``'a'``
format type of the C stdlib), that ``'#'`` option is used to control the
prefix (off by default).

Trailing zeros (and the dot) in fractional part are excluded.

Examples:
```pycon
>>> f'{-0.1:#x}'
'-0x1.999999999999ap-4'
>>> (-0.1).hex()
'-0x1.999999999999ap-4'
>>> f'{3.14159:+#X}'
'+0X1.921F9F01B866EP+1'
>>> f'{3.14159:.3x}'
'1.922p+1'
```

Minor changes:
* added Py_hexdigits_upper constant
* tests for RaisingNumber are moved (to test also bytes)

vstinner commented on Doc/library/stdtypes.rst:
```diff
@@ -2473,8 +2473,19 @@ Notes:
+   inserted before the first digit.
+
+   For floats, represent the number by a hexadecimal string in the style
+   ``[±]0xh.[hhhp±d``, where there is one hexadecimal digit before the
```
There is a missing `]`, no? Maybe:

```suggestion
   ``[±]0xh.[hhh]p±d``, where there is one hexadecimal digit before the
```

vstinner commented on Lib/test/test_float.py:
```diff
@@ -700,12 +700,20 @@ def test_format(self):
 
+        # hexadecimal format
+        self.assertEqual(format(float.fromhex('0x0.0030p+0'), 'x'), '0x1.8p-11')
+        self.assertEqual(format(float.fromhex('0x0.0040p+0'), 'x'), '0x1p-10')
```
I suggest adding `num = float.fromhex('0x0.0040p+0')` and then use *num*, rather than copy/paste it on each line.

vstinner commented on Python/pystrtod.c:
```diff
@@ -918,6 +932,97 @@ char * PyOS_double_to_string(double val,
+    assert(0 <= x && x < 16);
+    char c = Py_hexdigits[x];
+    if (upper) {
+        c = Py_TOUPPER(c);
```
That looks slow, it would be better to have an uppercase flavor of Py_hexdigits, such as Py_hexdigits_upper.

skirpichev replied:
In fact, this helper was literally moved from the ``floatobject.c``.

skirpichev replied:
@vstinner, I've added ``Py_hexdigits_upper``.  But I'm not sure if it worth (maybe as a local constant?): e.g. there are no other use cases across the codebase.

vstinner reviewed (commented):
I would prefer to not add `0x` prefix by default, and only add it if the alternative form (`#x`) is used.

skirpichev committed:
Address review: fix asserts

skirpichev committed:
Merge branch 'main' into hex-format

skirpichev committed:
Merge branch 'main' into hex-format

skirpichev committed:
Merge branch 'main' into hex-format

skirpichev committed:
Expand news entry & change section

skirpichev committed:
Merge branch 'master' into hex-format

skirpichev committed:
Don't clear trailing zeros + misc for float.hex() compatibility

- special workaround for zero with automatic precision
- revert back hex() tests
- adapt new tests

skirpichev committed:
_Py_dg_dtoa_hex() got special option for float.hex()

skirpichev committed:
Obviously, this is not going in 3.13...

vstinner commented on Lib/test/test_complex.py:
```diff
@@ -814,7 +814,7 @@ def test_format(self):
 
         # integer presentation types are an error
-        for t in 'bcdoxX':
+        for t in 'bcdo':
```
It seems like you added support for "x" in complex numbers. Would it be possible to not add it?

skirpichev replied:
I think so, but why we want this?

vstinner reviewed (commented):
I would prefer to not support complex, not add `x` format to `str % args` for float since it's different from C printf(), nor changing `bytes % args`.

I am fine with only changing `float.__format__()` to accept `x` and `X` formats with `#` alternative form.

skirpichev committed:
Merge branch 'main' into hex-format

skirpichev committed:
Merge branch 'main' into hex-format

skirpichev committed:
Merge branch 'master' into hex-format

skirpichev committed:
Address review: revert changes in old-style formatting

skirpichev committed:
Address review: revert x/X formatting for complexes

vstinner commented on Doc/c-api/conversion.rst:
```diff
@@ -151,6 +151,9 @@ The following functions provide locale-independent string to number conversions.
    .. versionadded:: 3.1
 
+.. versionchanged:: 3.14
+   Support ``'x'`` and ``'X'`` format types for :class:`float`.
```
```suggestion
   .. versionchanged:: 3.14
      Support ``'x'`` and ``'X'`` format types for :class:`float`.
```

vstinner commented on Doc/library/string.rst:
```diff
@@ -588,6 +588,30 @@ The available presentation types for :class:`float` and
+   |         | If the ``'#'`` option is specified, the prefix ``'0x'``  |
+   |         | will be inserted before an integer part.                 |
+   +---------+----------------------------------------------------------+
+   | ``'X'`` | Same as ``'x'``, but uses uppercase digits, the ``0X``   |
```
I don't think that digits can be upper or lower cases:

```suggestion
   | ``'X'`` | Same as ``'x'``, but uses uppercase, the ``0X``   |
```

skirpichev replied:
Hexadecimal digits can:)  I thought this was obvious from the context, so "hexadecimal" was omitted (doesn't fit to one line).  I can add this.

I worry that in your version it might be not clear, that uppercase used not just for the prefix.

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+   use_alt_formatting - nonzero if the hexadecimal prefix should be added.
+   upper - nonzero, if uppercase letters should be used for hexadecimal
+           numbers, prefix and the exponent separator.
+   _hex - private flag to keep float.hex() behaviour
```
Please describe the effect of the option rather than just saying "private flag".

I suggest to rename the parameter to "float_hex" and describe it as "use float.hex() format".

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+    double m = frexp(fabs(x), &e);
+
+    if (autoprec) {
+        precision = (DBL_MANT_DIG + 2 - (DBL_MANT_DIG+2)%4)/4;
```
Can you add a comment explaining this formula?

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+    int e, autoprec = precision < 0;
+    double m = frexp(fabs(x), &e);
+
+    if (autoprec) {
```
nitpick: can you declare autoprec variable just before this if?

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+            double frac = ldexp(m, 4*precision);
+            frac -= floor(frac);
+            frac *= 16.0;
+            m += ldexp(frac >= 8.0, -4*precision);
```
`frac >= 8.0` result is 1 or 0? Maybe write it more explicitly?

```suggestion
            if (frac >= 8.0) {
                m += ldexp(1.0, -4*precision);
            }
```

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
-   of the form 4k+1. */
-#define TOHEX_NBITS DBL_MANT_DIG + 3 - (DBL_MANT_DIG+2)%4
+    /* Allocate space for [±][0x]  h[.] [hhhhhhhh]   p±  d            '\0' */
+    char *s = PyMem_Malloc(1 + 2 + 2 +   precision + 2 + DBL_MAX_EXP + 1);
```
Can you only add +2 for "0x" if use_alt_formatting? Something like:

```
size = ...;
if (use_alt_formatting) size += 2;  // "0x"
```

Would it make sense to use alloca() since the string is quite short? What is the maximum size?

skirpichev replied:
> What is the maximum size?

Well, for typical 64-bit floats:
```pycon
>>> (sys.float_info.mant_dig + 2 - (sys.float_info.mant_dig+2)%4)//4
13
```
Given that exp length will be <= 4 - we can estimate size as ~25.

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+    si++;
+    si += snprintf(s + si, DBL_MAX_EXP + 1, "%+d", e) + 1;
+
+    return PyMem_Realloc(s, si);
```
Realloc can fail (even if it's unlikely), and this code leaks memory in this case.

Is it really needed to resize? Just remove the call no? The buffer is going to be freed quickly anyway.

vstinner commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
+    s[si] = upper ? 'P' : 'p';
+    si++;
+    si += snprintf(s + si, DBL_MAX_EXP + 1, "%+d", e) + 1;
+
```
Does this code always write a null character? You should use +2 to write also the null character, no?

skirpichev replied:
Yep, fixed.  And exponent size was overestimated.

skirpichev committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

skirpichev committed:
address review: rename/document _hex option

skirpichev commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,118 @@ hex_from_char(char c) {
+    int e, autoprec = precision < 0;
+    double m = frexp(fabs(x), &e);
+
+    if (autoprec) {
```
```suggestion
    int e;
    double m = frexp(fabs(x), &e);
    int autoprec = precision < 0;

    if (autoprec) {
```
Do you mean this?  // per PEP7 "blank line after local variables declaration"

vstinner replied:
I mean grouping e and m, empty line, declare autoprec and useautoproc.

> Do you mean this? // per PEP7 "blank line after local variables declaration"

Python uses C11, we can declare variables in the middle of the function. I prefer to defer variable declaration until it's used (just before its first usage/assignment), to reduce their scope, and so ease review/maintenance (less variables to think about when reviewing).

skirpichev committed:
Merge branch 'master' into hex-format

skirpichev committed:
address review: move autoprec decl & comment on tohex_nbits const

picnixz commented on Objects/floatobject.c:
```diff
@@ -1208,11 +1201,118 @@ hex_from_char(char c) {
+            /* for compatibility with float.hex(), we keep just one
+               digit of zero */
+            precision = 1;
+        }
```
```suggestion
        if (!x && _hex) {
            /* for compatibility with float.hex(), we keep just one
               digit of zero */
            precision = 1;
        } 
        else {
	        precision = (DBL_MANT_DIG + 2 - (DBL_MANT_DIG + 2) % 4) / 4;
        }
```

skirpichev replied:
No, thanks.  I hope eventually if statement can be removed.

mdickinson replied:
> No, thanks. I hope eventually if statement can be removed.

Why would you hope that?

picnixz reviewed (commented):
There are some nitpicks that I only mention, namely "4*precision" -> "4 * precision" but I'm not sure whether you want to put spaces around simple operators like that.

skirpichev committed:
address review: memory allocation

* improve:) exp estimation
* no Realloc
* dynamic size

skirpichev committed:
fix: catch PyMem_Malloc() error and document returned value

skirpichev committed:
fix: warnings

mdickinson commented on Include/internal/pycore_floatobject.h:
```diff
@@ -55,6 +55,8 @@ extern PyObject* _Py_string_to_number_with_underscores(
 
 extern double _Py_parse_inf_or_nan(const char *p, char **endptr);
 
+extern char * _Py_dg_dtoa_hex(double x, int precision, int always_add_sign,
```
I suggest calling this something clearer like `_Py_float_to_hex`. The `dg_dtoa` stuff doesn't make much sense here: "dg" is a reference to David Gay, since he wrote the original version of the dtoa.c code that Python's dtoa.c is baed on; this code has nothing to do with Gay's code.

skirpichev replied:
Thanks, does make sense.

Initially that helper was a static function in pystrtod.c; just as ``_Py_dg_dtoa()`` it, well, "convert double to ASCII string"...  This was exported and moved to floatobject.c to (1) slighly reduce diff and (2) support workaround for hex().

I think (1) goal wasn't very successful.  Then, maybe this helper belongs to pystrtod.c?  If so, ``_Py_dtoa_hex()`` sounds ok?

skirpichev committed:
address review: _Py_dg_dtoa_hex -> _Py_float_to_hex

skirpichev committed:
address review: round-to-even

skirpichev committed:
Merge branch 'master' into hex-format

skirpichev committed:
+

skirpichev committed:
Merge branch 'master' into hex-format-test

skirpichev committed:
cleanup

skirpichev committed:
+1
